tests/Unit/Evolution/Systems/ScalarWave/Test_Characteristics.cpp

```diff
@@ -0,0 +1,380 @@
+
+  using Affine = domain::CoordinateMaps::Affine;
+  using Affine3D =
+      domain::CoordinateMaps::ProductOf3Maps<Affine, Affine, Affine>;
+  const auto coord_map =
+      domain::make_coordinate_map<Frame::Logical, Frame::Inertial>(Affine3D{
+          Affine{-1., 1., lower_bound[0], upper_bound[0]},
+          Affine{-1., 1., lower_bound[1], upper_bound[1]},
+          Affine{-1., 1., lower_bound[2], upper_bound[2]},
+      });
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(mesh);
+  const auto x = coord_map(x_logical);
+  // Arbitrary time for time-dependent solution.
+  const double t = 0.;
+
+  // Evaluate analytic solution
+  const auto vars = solution.variables(
+      x, t,
+      tmpl::list<ScalarWave::Pi, ScalarWave::Phi<Dim>, ScalarWave::Psi>{});
+  // Get ingredients
+  const size_t n_pts = x.begin()->size();
+  const auto gamma_2 = make_with_value<Scalar<DataVector>>(x, 0.1);
+  const auto& psi = get<ScalarWave::Psi>(vars);
+  const auto& phi = get<ScalarWave::Phi<Dim>>(vars);
+  const auto& pi = get<ScalarWave::Pi>(vars);
+  // Outward 3-normal to the surface on which characteristic fields are needed
+  auto unit_normal_one_form =
+      make_with_value<tnsr::i<DataVector, Dim, Frame::Inertial>>(x, 1.);
+  const auto norm_of_one_form = magnitude(unit_normal_one_form);
+  for (size_t i = 0; i < Dim; ++i) {
+    unit_normal_one_form.get(i) /= get(norm_of_one_form);
+  }
+
+  // Fundamental fields (psi, pi, phi) have already been computed locally.
+  // Now, check that these locally computed fields match returned ones
+  // First, compute characteristic fields locally
+  Scalar<DataVector> phi_dot_normal{DataVector(n_pts, 0.)};
+  for (size_t i = 0; i < Dim; ++i) {
+    get(phi_dot_normal) += unit_normal_one_form.get(i) * phi.get(i);
+  }
+
+  tnsr::i<DataVector, Dim, Frame::Inertial> phi_dot_projection_tensor{
+      DataVector(n_pts, 0.)};
+  for (size_t i = 0; i < Dim; ++i) {
+    phi_dot_projection_tensor.get(i) =
+        phi.get(i) - unit_normal_one_form.get(i) * get(phi_dot_normal);
+  }
+
+  const auto& upsi = psi;
+  const auto& uzero = phi_dot_projection_tensor;
+  const Scalar<DataVector> uplus{get(pi) + get(phi_dot_normal) -
+                                 get(gamma_2) * get(psi)};
+  const Scalar<DataVector> uminus{get(pi) - get(phi_dot_normal) -
+                                  get(gamma_2) * get(psi)};
+  // Second, obtain fundamental fields using compute tag
+  const auto ffields =
+      ScalarWave::EvolvedFieldsFromCharacteristicFieldsCompute<Dim>::function(
+          gamma_2, upsi, uzero, uplus, uminus, unit_normal_one_form);
+  const auto& psi_from_func = get<ScalarWave::Psi>(ffields);
+  const auto& pi_from_func = get<ScalarWave::Pi>(ffields);
+  const auto& phi_from_func = get<ScalarWave::Phi<Dim>>(ffields);
+
+  CHECK_ITERABLE_APPROX(psi, psi_from_func);
+  CHECK_ITERABLE_APPROX(pi, pi_from_func);
+  CHECK_ITERABLE_APPROX(phi, phi_from_func);
```

Comment:
it would also be a good check to test that you get the same back if you computed the char vars using the function you're adding.

Reply:
added